bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ClientLibrariesImpl.java

```diff
@@ -263,23 +302,58 @@ private String getInline(LibraryType libraryType) {
         return categories;
     }
 
-    private void addClientLibraries(Resource componentRes, Collection<ClientLibrary> libraries) {
-        if (componentRes == null) {
+    /**
+     * Adds client libraries to the provided collection, starting for the resource type.
+     *
+     * @param resourceType - the resource type to look into for client libraries
+     * @param libraries - the provided collection of libraries to add to
+     * @param seenResourceTypes - a set of resource types that were previously searched into, to avoid inheritance loops
+     */
+    private void addClientLibraries(String resourceType, Collection<ClientLibrary> libraries, Set<String> seenResourceTypes) {
```

Comment:
I would extract the logic to retrieve the inherited resource types and move it to Utils, similar to the logic to retrieve the resource types of the page and of the XF components.
I created a PR as a proposal: https://github.com/adobe/aem-core-wcm-components/pull/1101

Reply:
The same refactoring could also be applied to `ComponentFilesImpl`.

Reply:
For `ComponentFiles` we might not want to walk the inheritance tree and collect everything. For the AMP use-case we would want to be able to override the `customheadlib.amp.html` from the super-type.

Reply:
I added a stop flag to allow file overrides in a32ae26 and added some test cases for this, including multiple level inheritance.

Reply:
Included super-types logic from #1101 in 35531fb